fix(rocprofiler-sdk): [rocpd] Add new counter and sampling tracks to perfetto

amd-atumakae opened the pull request:
# PR Details

## Associated Jira Ticket Number/Link
JIRA ID: SWDEV-557789 and JIRA ID: ROCPDSNA-53

## What type of PR is this? (check all applicable)

- [ ] Refactor
- [ ] Feature
- [x] Bug Fix
- [ ] Optimization
- [ ] Documentation Update
- [ ] Continuous Integration

## Technical details

This PR allows ROCPD to convert CPU & GPU samples from a rocprof-sys database.  In the JIRA tickets, there are sample databases attached to test the conversion.  

You will see CPU samples and GPU samples now plotted correctly.
You will see flow arrows drawn for kernel dispatches correctly.  None should be drawn for any sampling.

## Added/updated tests?

- [ ] Yes
- [ ] No, Does not apply to this PR.

## Updated CHANGELOG?

<!-- Needed for Release updates for a ROCm release. -->

- [ ] Yes
- [ ] No, Does not apply to this PR.

## Added/Updated documentation?

- [ ] Yes
- [ ] No, Does not apply to this PR.

amd-atumakae committed:
Add thread counter tracks

amd-atumakae committed:
Add thread counter tracks

amd-atumakae committed:
Add missing GPU and CPU counter tracks

amd-atumakae committed:
Add missing GPU and CPU counter tracks

amd-atumakae committed:
Query optimisations

amd-atumakae committed:
Query optimisations

yhuiYH commented on projects/rocprofiler-sdk/source/lib/python/rocpd/libpyrocpd.cpp:
```diff
@@ -469,12 +468,151 @@ PYBIND11_MODULE(libpyrocpd, pyrocpd)
+                        const std::string create_samples_view =
+                            "CREATE TEMP VIEW "
+                            "    `samples_temp` AS "
+                            "SELECT "
```
```suggestion
                        const std::string create_samples_view = R"(
                            CREATE TEMP VIEW
                                `samples_temp` AS
                            SELECT
                            ...)";
```
Can you use R"(...)" instead of these multiple quotes?

amd-atumakae replied:
Done

yhuiYH commented on projects/rocprofiler-sdk/source/lib/python/rocpd/libpyrocpd.cpp:
```diff
@@ -469,12 +468,151 @@ PYBIND11_MODULE(libpyrocpd, pyrocpd)
+
+                        sqlite3_stmt* _stmt_samples_temp;
+                        uint64_t      prepare_samples_status = sqlite3_prepare_v2(
+                            conn, create_samples_view.c_str(), -1, &_stmt_samples_temp, nullptr);
```
Instead of using sqlite3_prepare_v2() directly, you can probably use the macro `execute_raw_sql_statements` like https://github.com/ROCm/rocm-systems/blob/develop/projects/rocprofiler-sdk/source/lib/output/generateRocpd.cpp#L615

The SQL helper functions are in this folder: https://github.com/ROCm/rocm-systems/blob/develop/projects/rocprofiler-sdk/source/lib/output/sql/common.hpp#L125

amd-atumakae replied:
Done

yhuiYH commented on projects/rocprofiler-sdk/source/lib/python/rocpd/libpyrocpd.cpp:
```diff
@@ -469,12 +468,151 @@ PYBIND11_MODULE(libpyrocpd, pyrocpd)
 
+                        const std::string create_samples_view =
+                            "CREATE TEMP VIEW "
+                            "    `samples_temp` AS "
```
```suggestion
                            "    `samples_schema_3_0` AS "
```
Use a better name instead of "samples_temp".  Since the next schema will be version bumped, we can refer to the 3.0 schema for these.

amd-atumakae replied:
Done

yhuiYH commented on projects/rocprofiler-sdk/source/lib/python/rocpd/libpyrocpd.cpp:
```diff
@@ -469,12 +468,151 @@ PYBIND11_MODULE(libpyrocpd, pyrocpd)
+
+                        const std::string create_pmc_event_view =
+                            "CREATE TEMP VIEW "
+                            "    `pmc_events_temp` AS "
```
```suggestion
                            "    `pmc_events_temp` AS "
```
Don't use temp... maybe refer to "schema_3_0" instead

amd-atumakae replied:
Done

yhuiYH commented on projects/rocprofiler-sdk/source/lib/python/rocpd/libpyrocpd.cpp:
```diff
@@ -469,12 +468,151 @@ PYBIND11_MODULE(libpyrocpd, pyrocpd)
+                        auto pmc_info = rocpd::sql_generator<rocpd::types::pmc_info>{
+                            conn, select_guid_nid_pid("pmc_info")};
+
+                        const std::string create_pmc_event_view =
```
Switch to R"(...)"; style string

amd-atumakae replied:
Done

yhuiYH commented on projects/rocprofiler-sdk/source/lib/python/rocpd/source/perfetto.cpp:
```diff
@@ -379,13 +380,21 @@ write_perfetto(
                 auto& track = thread_tracks.at(itr.tid);
                 auto  _name = itr.name;
 
+                /* Temporary crash fix
```
Need to remove this comment later....

amd-atumakae replied:
Removed the fix, not needed anymore.

yhuiYH commented on projects/rocprofiler-sdk/source/lib/python/rocpd/source/perfetto.cpp:
```diff
@@ -379,13 +380,21 @@ write_perfetto(
                 }
+                */
+
+                const char* category_name = get_category_string(itr.category);
```
In Regions, I think we need to separate TRACE data from the TIMER_SAMPLING data tracks.  The TIMER_SAMPLING values have their own category.  We should create timer_sampling tracks for those, similar to the native Perfetto trace had.

Oh, and with timer_sampling tracks, we don't populate any flow arrows.

amd-atumakae replied:
Done

amd-atumakae committed:
Udpate SQL queries

amd-atumakae committed:
Udpate SQL queries

amd-atumakae committed:
Remove temporary crash fix

amd-atumakae committed:
Remove temporary crash fix

amd-atumakae committed:
Move timer sampling data to separate tracks

amd-atumakae committed:
Move timer sampling data to separate tracks

amd-atumakae committed:
Fix format

amd-atumakae committed:
Fix format

amd-atumakae committed:
Create temp views one time

amd-atumakae committed:
Create temp views one time

yhuiYH commented on projects/rocprofiler-sdk/source/lib/python/rocpd/libpyrocpd.cpp:
```diff
@@ -419,8 +418,112 @@ PYBIND11_MODULE(libpyrocpd, pyrocpd)
+                auto* conn = rocpd::interop::get_connection(std::move(obj));
+
+                const std::string create_samples_view = R"(
+                            CREATE TEMP VIEW
```
```suggestion
                            CREATE TEMP VIEW IF NOT EXISTS
```
When trying to convert 2 databases, this errors out because the view already exits.  Same with the next view.

amd-atumakae replied:
Done

yhuiYH commented on projects/rocprofiler-sdk/source/lib/python/rocpd/libpyrocpd.cpp:
```diff
@@ -419,8 +418,112 @@ PYBIND11_MODULE(libpyrocpd, pyrocpd)
+                execute_raw_sql_statements(conn, create_samples_view);
+
+                const std::string create_pmc_event_view = R"(
+                            CREATE TEMP VIEW
```
```suggestion
                            CREATE TEMP VIEW IF NOT EXISTS
```
Same fix for multiple DBs required.

amd-atumakae replied:
Done

yhuiYH commented on projects/rocprofiler-sdk/source/lib/python/rocpd/libpyrocpd.cpp:
```diff
@@ -419,8 +418,112 @@ PYBIND11_MODULE(libpyrocpd, pyrocpd)
+                                AND TH.guid = T.guid;
+                        )";
+
+                execute_raw_sql_statements(conn, create_samples_view);
```
We should only create these views, if the input databases are schema 3.0

yhuiYH replied:
We will add provisioning for the schema 3 check in future iterations.

yhuiYH commented on projects/rocprofiler-sdk/source/lib/python/rocpd/source/perfetto.cpp:
```diff
@@ -389,12 +434,18 @@ write_perfetto(
                                   track,
                                   itr.start,
-                                  ::perfetto::Flow::Global(itr.stack_id ^ uuid_pid),
+                                  ::perfetto::Flow::Global(itr.stack_id ^ this_pid_track.uuid),
```
Still seeing a lot of `flow` arrows in the timer_sampling data.  Can we get rid of those?

amd-atumakae replied:
Fixed

yhuiYH reviewed (dismissed):
Can you try to convert 2 rocprofv3 databases?  The perfetto file only has 1 trace now, so these changes broke the existing logic to aggregate multiple DBs into 1 perfetto file.

We should always test aggregating multiple databases too.

amd-atumakae committed:
Remove flow arrows

amd-atumakae committed:
Remove flow arrows

amd-atumakae committed:
Fix for multiple databases

amd-atumakae committed:
Fix for multiple databases

yhuiYH commented on projects/rocprofiler-sdk/source/lib/python/rocpd/source/perfetto.cpp:
```diff
@@ -281,6 +283,13 @@ write_perfetto(
 
             thread_tracks.emplace(itr.tid, _track);
+
+            auto _sampling_track = ::perfetto::Track::Global(++global_track_index);
```
```suggestion
            auto _sampling_track = ::perfetto::Track{get_hash_id(fmt::format("sampling_{}", itr.tid)), this_pid_track};
```
Instead of using global_track_index, I used a hash, and we can get rid of the "Global Track Events" section in the perfetto.

amd-atumakae replied:
Done.

yhuiYH commented on projects/rocprofiler-sdk/source/lib/python/rocpd/source/perfetto.cpp:
```diff
@@ -300,7 +309,7 @@ write_perfetto(
                 _namess << "(UNK)";
 
-            auto _track = ::perfetto::Track{get_hash_id(_namess.str()), this_pid_track};
+            auto _track = ::perfetto::Track::Global(++global_track_index);
```
```suggestion
            auto _track = ::perfetto::Track{get_hash_id(_namess.str()), this_pid_track};
```
Don't really like the global_track_index.  Can we go back to the original?

amd-atumakae replied:
Done.

yhuiYH commented on projects/rocprofiler-sdk/source/lib/python/rocpd/source/perfetto.cpp:
```diff
@@ -177,27 +177,28 @@ write_perfetto(
 
-    auto uuid_pid       = common::fnv1a_hasher::combine(this_nid, this_pid_init_ns, this_pid);
-    auto this_pid_track = ::perfetto::Track{uuid_pid, ::perfetto::Track{}};
+    auto this_pid_track = ::perfetto::Track::Global(++global_track_index);
```
```suggestion
    auto uuid_pid       = common::fnv1a_hasher::combine(this_nid, this_pid_init_ns, this_pid);
    auto this_pid_track = ::perfetto::Track{uuid_pid, ::perfetto::Track{}};
```
Can we go back to original?  If the objective is to just get "Sampling" tracks to create their own tracks, then just making a UUID unique should work?  I provided a suggestion in line 287.

amd-atumakae replied:
Done.

yhuiYH commented on projects/rocprofiler-sdk/source/lib/python/rocpd/source/perfetto.cpp:
```diff
@@ -389,12 +403,18 @@ write_perfetto(
                                   track,
                                   itr.start,
-                                  ::perfetto::Flow::Global(itr.stack_id ^ uuid_pid),
+                                  ::perfetto::Flow::Global(++global_flow_index),
```
```suggestion
                                  ::perfetto::Flow::Global((itr.stack_id == 0) ? (++global_flow_index) : (itr.stack_id ^ this_pid_track.uuid)),
```
It's not pretty, but I change it to this, so that IF a proper stack_id is provided, it will draw the flow arrows.  
BUT we need to get global_flow_index, to some high initial starting number.  So that it avoids collisions with actual stack_id values.

amd-atumakae replied:
Done.

yhuiYH commented on projects/rocprofiler-sdk/source/lib/python/rocpd/source/perfetto.cpp:
```diff
@@ -177,27 +177,28 @@ write_perfetto(
+    static auto     orig_process_track = ::perfetto::ProcessTrack::Current();
+    static auto     orig_process_desc  = orig_process_track.Serialize();
+    static uint64_t global_flow_index  = 0;
+    static uint64_t global_track_index = 0;
```
```suggestion
    static uint64_t global_flow_index  = 0x100000000;
```
I set it to some high value, just for the flow arrows

amd-atumakae replied:
Done.

amd-atumakae committed:
Revert changes for track indexing

amd-atumakae committed:
Revert changes for track indexing

amd-atumakae committed:
Improved performance when multiple databases are attached

amd-atumakae committed:
Improved performance when multiple databases are attached

amd-atumakae committed:
Fix flow indexes

amd-atumakae committed:
Fix flow indexes

amd-atumakae committed:
Fix perfetto validation

amd-atumakae committed:
Fix perfetto validation

yhuiYH committed:
Fix thread_id, don't decrement it in Track title

yhuiYH committed:
Fix thread_id, don't decrement it in Track title

yhuiYH committed:
Use thread index as CPU Sample Tracks label instead of thread id, to match rocprof-sys

yhuiYH committed:
Use thread index as CPU Sample Tracks label instead of thread id, to match rocprof-sys

yhuiYH committed:
fix style changes and simplify with get_flow_index

yhuiYH committed:
fix style changes and simplify with get_flow_index

yhuiYH committed:
apply some copilot suggested fixes

yhuiYH committed:
apply some copilot suggested fixes

yhuiYH commented on projects/rocprofiler-sdk/source/lib/python/rocpd/source/perfetto.cpp:
```diff
@@ -789,6 +880,42 @@ write_perfetto(
+                tracing_session->FlushBlocking();
+            }
+        }
+
```
Done

yhuiYH commented on projects/rocprofiler-sdk/source/lib/python/rocpd/libpyrocpd.cpp:
```diff
@@ -423,6 +422,129 @@ PYBIND11_MODULE(libpyrocpd, pyrocpd)
+                            FROM
+                                `rocpd_info_pmc` PMC_I
+                                INNER JOIN `rocpd_info_agent` A ON A.id = PMC_I.agent_id
+                                AND A.guid = PMC_I.guid
```
Good catch, fixed semi-colon. But the closing bracket is wrong.

yhuiYH commented on projects/rocprofiler-sdk/source/lib/python/rocpd/libpyrocpd.cpp:
```diff
@@ -423,6 +422,129 @@ PYBIND11_MODULE(libpyrocpd, pyrocpd)
+                                AND PMC_I.guid = T.guid
+                                LEFT OUTER JOIN `rocpd_info_thread` TH ON TH.id = T.tid
+                                AND TH.guid = T.guid;
+                        )";
```
Nope, brackets are fine.

yhuiYH commented on projects/rocprofiler-sdk/source/lib/python/rocpd/libpyrocpd.cpp:
```diff
@@ -423,6 +422,129 @@ PYBIND11_MODULE(libpyrocpd, pyrocpd)
+                                `rocpd_pmc_event` PMC_E
+                                INNER JOIN `rocpd_info_pmc` PMC_I ON PMC_I.id = PMC_E.pmc_id
+                                AND PMC_I.guid = PMC_E.guid;
+                        )";
```
Nope, brackets are fine.

yhuiYH commented on projects/rocprofiler-sdk/source/lib/python/rocpd/libpyrocpd.cpp:
```diff
@@ -473,11 +595,17 @@ PYBIND11_MODULE(libpyrocpd, pyrocpd)
                             conn, select_guid_nid_pid("threads")};
 
+                        auto pmc_info = rocpd::sql_generator<rocpd::types::pmc_info>{
+                            conn, select_guid_nid_pid("pmc_info")};
```
Good, removed unused code.

yhuiYH commented on projects/rocprofiler-sdk/source/lib/python/rocpd/source/perfetto.cpp:
```diff
@@ -507,12 +558,39 @@ write_perfetto(
+                    // For GPU tracks, use (0, agent_abs_index) as the track index
+                    std::pair<uint32_t, uint32_t> track_index = {itr.thread_index,
+                                                                 itr.agent_abs_index};
+                    auto& track_data = sample_counter_tracks[itr.symbol][track_index];
```
Yup, just avoid case and set to <uint64_t>

yhuiYH commented on projects/rocprofiler-sdk/source/lib/python/rocpd/source/perfetto.cpp:
```diff
@@ -507,12 +558,39 @@ write_perfetto(
+                    track_data.unit_name   = itr.units;
+                    track_data.description = itr.short_description;
+                    track_data.samples.emplace(itr.timestamp,
+                                               pmc_event_values[itr.symbol][itr.event_id]);
```
This should be ok.

yhuiYH commented on projects/rocprofiler-sdk/source/lib/python/rocpd/libpyrocpd.cpp:
```diff
@@ -402,12 +402,11 @@ PYBIND11_MODULE(libpyrocpd, pyrocpd)
+                auto ret_index = tool::create_agent_index(output_cfg.agent_index_value,
+                                                          _agent.node_id,        // absolute index
+                                                          _agent.logical_index,  // relative index
+                                                          _agent.type_index,  // type-relative index
```
changed

yhuiYH committed:
simplify get_category code in perfetto

yhuiYH committed:
simplify get_category code in perfetto

yhuiYH committed:
Make adjustments for latest rocprof-sys changes on how they stores samples
- Remove unit conversion workarounds now that rocprof-sys fixed the stored values to match the units
- Need to add more agent indexes to the views
- Process sample data for correct agent number only

yhuiYH committed:
Make adjustments for latest rocprof-sys changes on how they stores samples
- Remove unit conversion workarounds now that rocprof-sys fixed the stored values to match the units
- Need to add more agent indexes to the views
- Process sample data for correct agent number only

srihari-uttanur commented on projects/rocprofiler-sdk/source/lib/python/rocpd/source/perfetto.cpp:
```diff
@@ -522,12 +570,43 @@ write_perfetto(
+                                                                 itr.agent_type_index};
+                    auto& track_data = sample_counter_tracks[itr.symbol][track_index];
+
+                    const auto _agent      = agent_data.at(itr.agent_abs_index).first;
```
agent_data.at(..) can throw, its better to use find and log a warning for better error handling.

yhuiYH replied:
agent_data.at(..) is used throughout the conversion.  Ack this is possible (but highly unlikely), but to change, would require significant rework of this file.  Will park this.

srihari-uttanur commented on projects/rocprofiler-sdk/source/lib/python/rocpd/source/perfetto.cpp:
```diff
@@ -804,6 +890,41 @@ write_perfetto(
+                if(track_data.size() != 1)
+                {
+                    if(data.agent_type == "CPU")
+                        track_name << " [" << track_index.first - 1 << "]";
```
Possible unsigned underflow when index is 0.

yhuiYH replied:
Sure, error checked this.  Thx.

srihari-uttanur commented on projects/rocprofiler-sdk/source/lib/python/rocpd/source/perfetto.cpp:
```diff
@@ -415,25 +433,53 @@ write_perfetto(
-        };
+        for(const auto& itr : pmc_event_gen.get(ditr))
+        {
+            pmc_event_values[itr.symbol][itr.event_id][itr.agent_abs_index] = itr.value;
```
Are we aggregating pmc events here ? There can be multiple rows with same symbol/event_id and agent_abs_index ?

yhuiYH replied:
No, there should only be 1 row with same symbol/event_id, otherwise the data is incorrect.  The perfetto code, should just convert what is stored in the data, which is take the value it find.  No need for accumulation.

srihari-uttanur commented on projects/rocprofiler-sdk/source/lib/python/rocpd/source/perfetto.cpp:
```diff
@@ -522,12 +570,43 @@ write_perfetto(
+                if(!itr.symbol.empty())
+                {
+                    // For CPU tracks, use (thread_index, 0) as the track index
+                    // For GPU tracks, use (0, agent_type_index) as the track index
```
can you elaborate about this comment ?

yhuiYH replied:
Elaborated a bit in the comment:
`// Collect index for the track name in perfetto (ie, CPU [0], GPU [0], GPU [1])`

yhuiYH committed:
Add changelog.md entry and make changes for review comments

yhuiYH committed:
Add changelog.md entry and make changes for review comments

jrmadsen commented on projects/rocprofiler-sdk/source/lib/python/rocpd/libpyrocpd.cpp:
```diff
@@ -423,6 +422,127 @@ PYBIND11_MODULE(libpyrocpd, pyrocpd)
+                                    WHEN 'thread_cpu_time' THEN 'Thread CPU Time'
+                                    WHEN 'thread_page_fault' THEN 'Thread Page Faults'
+                                    WHEN 'thread_peak_memory' THEN 'Thread Peak Memory Usage'
+                                    ELSE PMC_I.description
```
This looks very hardcoded. I don't like it. Are there alternatives?

yhuiYH replied:
OK, broke that long SQL string so that it generates the `units` and `short_description` fields.  If we need to add new items to those, we just add to the c++ mapping tables, instead of changing the strings in the query.  https://github.com/ROCm/rocm-systems/pull/1324/commits/f30e8a99d581e3c38907ef1743ed3b1941e2e5c3

yhuiYH committed:
break info_pmc_schema_3_0 table to generate the sql string from mapping pairs, avoid hard-coded sql

yhuiYH committed:
break info_pmc_schema_3_0 table to generate the sql string from mapping pairs, avoid hard-coded sql

yhuiYH committed:
change to use string.append for clang-tidy fix

yhuiYH committed:
change to use string.append for clang-tidy fix

amd-justchen commented:
Noticing that these tests are not running:

> pending checks

  [rocprofiler-sdk Continuous Integration / AddressSanitizer • mi325 • ubuntu-22.04 (pull_request)](https://github.com/ROCm/rocm-systems/actions/runs/24900945610/job/72918169981?pr=1324)
  Queued — Waiting to run this check...
  [rocprofiler-sdk Continuous Integration / LeakSanitizer • mi325 • ubuntu-22.04 (pull_request)](https://github.com/ROCm/rocm-systems/actions/runs/24900945610/job/72918169990?pr=1324)
  Queued — Waiting to run this check...
  [rocprofiler-sdk Continuous Integration / ThreadSanitizer • mi325 • ubuntu-22.04 (pull_request)](https://github.com/ROCm/rocm-systems/actions/runs/24900945610/job/72918169978?pr=1324)
  Queued — Waiting to run this check...
  [rocprofiler-sdk Continuous Integration / UndefinedBehaviorSanitizer • mi325 • ubuntu-22.04 (pull_request)](https://github.com/ROCm/rocm-systems/actions/runs/24900945610/job/72918169968?pr=1324)
  Queued — Waiting to run this check...

They're waiting for `Requested labels: linux-mi325-8gpu-ossci-rocm-sandbox ` Which do not exist anymore
I'll cancel them for now, but I believe the workflow files need to be updated

yhuiYH committed:
add rocpd-sample-track tests which generates mock schema3 DB
- create_sample_db_schema3_0.py creates fake samples DB to simulate rocprof-sys data
- test converts this mock DB and validates that 2 tracks (one for CPU and another for GPU) exist
- to maintain rocpd backwards compatibility, other create_sample_db.py scripts would need to be created for newer schemas

yhuiYH committed:
add rocpd-sample-track tests which generates mock schema3 DB
- create_sample_db_schema3_0.py creates fake samples DB to simulate rocprof-sys data
- test converts this mock DB and validates that 2 tracks (one for CPU and another for GPU) exist
- to maintain rocpd backwards compatibility, other create_sample_db.py scripts would need to be created for newer schemas
